tests/cocotb/test_stream_mul.py

```diff
@@ -0,0 +1,309 @@
+    LOOP_COUNT_0 = 100
+    TEMPORAL_STRIDE_0 = 64
+    TEMPORAL_STRIDE_1 = 64
+    TEMPORAL_STRIDE_2 = 64
+    SPATIAL_STRIDE_0 = 8
+    SPATIAL_STRIDE_1 = 8
+    SPATIAL_STRIDE_2 = 8
+    BASE_PTR_0 = 0
+    BASE_PTR_1 = 32
+    BASE_PTR_2 = 64
+
+    # Start clock
+    clock = Clock(dut.clk_i, 10, units="ns")
+    cocotb.start_soon(clock.start())
+
+    # Initial reset values
+    # Need to clear for modelsim (or other simulator)
+    # Verilator assumes 0, no don't care states
+    dut.rst_ni.value = 0
+
+    # Always active assuming core can
+    # contiuously read data from streamer CSR
+    dut.io_csr_rsp_ready_i.value = 1
+
+    # Set DMA ports to 0 first before driving
+    dut.tcdm_dma_req_write_i.value = 0
+    dut.tcdm_dma_req_addr_i.value = 0
+    dut.tcdm_dma_req_data_i.value = 0
+    dut.tcdm_dma_req_strb_i.value = 0
+    dut.tcdm_dma_req_q_valid_i.value = 0
+
+    await snax_util.clock_and_wait(dut)
+
+    # Release reset
+    dut.rst_ni.value = 1
+
+    await snax_util.clock_and_wait(dut)
+
+    # Preload data into the TCDM subsys
+    # using the DMA ports
+    cocotb.log.info("Preload data with DMA control")
+
+    # Generate data to be processed
+    # Number of elements is dependent on:
+    # LOOP_COUNT_0 x (WIDE_DATA_WIDTH / NARROW_DATA_WIDTH)
+    narrow_golden_list = snax_util.gen_rand_int_list(
+        int(LOOP_COUNT_0 * (WIDE_DATA_WIDTH / NARROW_DATA_WIDTH)), MIN_VAL, MAX_VAL
+    )
+    wide_golden_list = snax_util.gen_wide_list(
+        narrow_golden_list, NARROW_DATA_WIDTH, WIDE_DATA_WIDTH
+    )
+
+    # Precompute golden results
+    narrow_golden_result = []
+
+    for i in range(LOOP_COUNT_0):
+        for j in range(SPATPAR):
+            temp_res = (
+                narrow_golden_list[i * SPATPAR * 2 + j]
+                * narrow_golden_list[(2 * i + 1) * SPATPAR + j]
+            )
+            temp_res = temp_res & (MAX_NARROW_VAL - 1)
+            narrow_golden_result.append(temp_res)
+
+    wide_golden_result = snax_util.gen_wide_list(
+        narrow_golden_result, NARROW_DATA_WIDTH, (NARROW_DATA_WIDTH * SPATPAR)
+    )
+
+    # Preload TCDM DMA subsys using DMA ports
+    wide_len = len(wide_golden_list)
+    for i in range(wide_len):
+        await snax_util.wide_tcdm_write(
+            dut, i * WIDE_BANK_INCREMENT, wide_golden_list[i]
+        )
+
+    await snax_util.wide_tcdm_clr(dut)
+
+    # Sanity check contents loaded into DMA
+    for i in range(wide_len):
+        tcdm_wide_val = await snax_util.wide_tcdm_read(dut, i * WIDE_BANK_INCREMENT)
+        snax_util.comp_and_assert(wide_golden_list[i], tcdm_wide_val)
+
+    await snax_util.wide_tcdm_clr(dut)
+
+    cocotb.log.info("Setting up of CSR registers and verifying if setup is correct")
+
+    # At this point we'll do explicit declartion
+    # of the tests so that it's understandable
+    # for other users
+
+    # Set number of iterations
+    await snax_util.reg_write(dut, CSR_LOOP_COUNT_0, LOOP_COUNT_0)
```

Comment:
Wait for the CSR setting I think it's fine. It just so happens that we set the same things in the same test. I think it would be better if it's separated for now.